ethexe/service/src/fast_sync.rs

```diff
@@ -0,0 +1,604 @@
+    }
+
+    fn is_empty(&self) -> bool {
+        self.buffered_requests.is_empty()
+    }
+
+    /// (total completed request, total pending requests)
+    fn stats(&self) -> (u64, u64) {
+        let completed = self.total_completed_requests;
+        let pending = self.total_pending_requests;
+        debug_assert!(completed <= pending, "{completed} <= {pending}");
+        (completed, pending)
+    }
+}
+
+impl Drop for RequestManager<'_> {
+    fn drop(&mut self) {
+        #[cfg(debug_assertions)]
+        {
+            let Self {
+                total_completed_requests,
+                total_pending_requests,
+                buffered_requests,
+                pending_requests,
+                responses,
+            } = self;
+            assert_eq!(total_completed_requests, total_pending_requests);
+            assert_eq!(*buffered_requests, HashMap::new());
+            assert_eq!(*pending_requests, HashMap::new());
+            assert_eq!(*responses, HashMap::new());
+        }
+    }
+}
+
+async fn sync_finalized_head(observer: &mut ObserverService) -> Result<H256> {
+    let highest_block = observer
+        .provider()
+        // we get finalized block to avoid block reorganization
+        // because we restore the database only for the latest block of a chain,
+        // and thus the reorganization can lead us to an empty block
+        .get_block(BlockId::Number(BlockNumberOrTag::Finalized))
+        .await
+        .context("failed to get latest block")?
+        .expect("latest block always exist");
+    let highest_block = H256(highest_block.header.hash.0);
+
+    log::info!("Syncing chain head {highest_block}");
+    observer.force_sync_block(highest_block).await?;
+    while let Some(event) = observer.next().await {
+        match event? {
+            ObserverEvent::Blob(_blob) => {
+                unreachable!("no blob events should occur before chain head is synced")
+            }
+            ObserverEvent::Block(_) => {}
+            ObserverEvent::BlockSynced(data) => {
+                debug_assert_eq!(highest_block, data.block_hash);
+                break;
+            }
+        }
+    }
+
+    Ok(highest_block)
+}
+
+async fn sync_from_network(
+    network: &mut NetworkService,
+    db: &Database,
+    program_states: &BTreeMap<ActorId, H256>,
+    latest_block_header: &BlockHeader,
+) -> Schedule {
+    let mut schedule_restorer = ScheduleRestorer::new(latest_block_header.height);
+
+    let program_states: HashMap<H256, HashSet<ActorId>> =
+        program_states
+            .iter()
+            .fold(HashMap::new(), |mut acc, (&program_id, &state)| {
+                acc.entry(state).or_default().insert(program_id);
+                acc
+            });
+    let mut manager = RequestManager::default();
+    for (&state, program_ids) in &program_states {
+        manager.add(state, RequestMetadata::ProgramState { program_ids });
+    }
```

Comment:
Looks like it will be easier to restore scheduler separatelly: after hash syncs, you can bypass all programs and restore scheduler for the actual state. These all requests with metadata looks tricky.